File: carkit/education/carkit_intro2av/carkit_intro2av/planning.py
#!/usr/bin/env python3

# CARKit learning annotation: implements the behavior described by this file's package and module.
"""Reference-shaped Intro2AV global-planning ROS node."""

import math
import time

from geometry_msgs.msg import PoseStamped
from nav_msgs.msg import OccupancyGrid, Odometry, Path
import rclpy
from rclpy.executors import ExternalShutdownException
from rclpy.node import Node
from rclpy.qos import (
    DurabilityPolicy,
    QoSProfile,
    ReliabilityPolicy,
    qos_profile_sensor_data,
)

from .planning_algorithm import PlanningConfig, compute_path


class Intro2AvPlanning(Node):
    """Own ROS state and delegate only path search to student code."""

    def __init__(self):
        super().__init__("carkit_intro2av_planning")
        self._declare_parameters()
        self.map_message = None
        self.odometry = None
        self.goal = None
        self.goal_revision = 0
        self.planned_revision = -1
        self.last_plan_time = 0.0

        self.plan_publisher = self.create_publisher(
            Path, self._parameter("plan_topic"), 10
        )
        map_qos = QoSProfile(
            depth=1,
            reliability=ReliabilityPolicy.RELIABLE,
            durability=DurabilityPolicy.TRANSIENT_LOCAL,
        )
        self.map_subscription = self.create_subscription(
            OccupancyGrid,
            self._parameter("map_topic"),
            self._on_map,
            map_qos,
        )
        self.odom_subscription = self.create_subscription(
            Odometry,
            self._parameter("odom_topic"),
            self._on_odometry,
            qos_profile_sensor_data,
        )
        self.goal_subscription = self.create_subscription(
            PoseStamped, self._parameter("goal_topic"), self._on_goal, 10
        )
        rate = max(0.1, float(self._parameter("planning_rate_hz")))
        self.timer = self.create_timer(1.0 / rate, self._update_plan)
        self.get_logger().info(
            "Intro2AV planner ready: map + odom + goal -> plan; "
            "implement planning_algorithm.compute_path"
        )

    def _declare_parameters(self):
        """Declare the ROS parameters and their safe default values."""
        defaults = {
            "map_topic": "/map",
            "odom_topic": "/odom",
            "goal_topic": "/goal_pose",
            "plan_topic": "/plan",
            "global_frame": "map",
            "planning_rate_hz": 2.0,
            "occupancy_threshold": 65,
            "allow_unknown": False,
            "inflation_radius_m": 0.25,
            "waypoint_spacing_m": 0.10,
            "goal_tolerance_m": 0.15,
        }
        for name, value in defaults.items():
            self.declare_parameter(name, value)

    def _parameter(self, name):
        """Return the current value of a declared ROS parameter."""
        return self.get_parameter(name).value

    def _on_map(self, message):
        self.map_message = message

    def _on_odometry(self, message):
        self.odometry = message

    def _on_goal(self, message):
        self.goal = message
        self.goal_revision += 1

    def _configuration(self):
        """Collect current parameters into an algorithm configuration."""
        return PlanningConfig(
            occupancy_threshold=int(self._parameter("occupancy_threshold")),
            allow_unknown=bool(self._parameter("allow_unknown")),
            inflation_radius_m=float(self._parameter("inflation_radius_m")),
            waypoint_spacing_m=float(self._parameter("waypoint_spacing_m")),
            goal_tolerance_m=float(self._parameter("goal_tolerance_m")),
        )

    def _update_plan(self):
        if self.map_message is None or self.odometry is None or self.goal is None:
            return
        now = time.monotonic()
        if (
            self.planned_revision == self.goal_revision
            and now - self.last_plan_time < 1.0
        ):
            return
        self.last_plan_time = now
        try:
            poses = compute_path(
                self.map_message,
                self.odometry,
                self.goal,
                self._configuration(),
            )
            path = self._validated_path(poses)
        except Exception as error:
            self.get_logger().error(f"Planning algorithm failed: {error}")
            path = self._empty_path()
        self.planned_revision = self.goal_revision
        self.plan_publisher.publish(path)

    def _empty_path(self):
        path = Path()
        path.header.stamp = self.get_clock().now().to_msg()
        path.header.frame_id = str(self._parameter("global_frame"))
        return path

    def _validated_path(self, poses):
        path = self._empty_path()
        for pose in poses:
            if not isinstance(pose, PoseStamped):
                raise TypeError("planner must return PoseStamped values")
            position = pose.pose.position
            if not math.isfinite(position.x) or not math.isfinite(position.y):
                raise ValueError("planner returned a non-finite waypoint")
            pose.header = path.header
            path.poses.append(pose)
        return path


def main(args=None):
    rclpy.init(args=args)
    node = Intro2AvPlanning()
    try:
        rclpy.spin(node)
    except (KeyboardInterrupt, ExternalShutdownException):
        pass
    finally:
        node.destroy_node()
        if rclpy.ok():
            rclpy.shutdown()

File: carkit/education/carkit_intro2av_cpp/src/planning.cpp
// Copyright 2026 CARKit maintainers
// Licensed under the Apache License, Version 2.0.

#include <algorithm>
#include <chrono>
#include <cmath>
#include <cstdint>
#include <memory>
#include <stdexcept>
#include <string>

#include "carkit_intro2av_cpp/planning_algorithm.hpp"
#include "geometry_msgs/msg/pose_stamped.hpp"
#include "nav_msgs/msg/occupancy_grid.hpp"
#include "nav_msgs/msg/odometry.hpp"
#include "nav_msgs/msg/path.hpp"
#include "rclcpp/rclcpp.hpp"

// CARKit learning annotation: implements the behavior described by this file's package and module.
using namespace std::chrono_literals;

class Intro2AvPlanning : public rclcpp::Node
{
public:
  Intro2AvPlanning() : Node("carkit_intro2av_cpp_planning")
  {
    map_topic_ = declare_parameter<std::string>("map_topic", "/map");
    odom_topic_ = declare_parameter<std::string>("odom_topic", "/odom");
    goal_topic_ = declare_parameter<std::string>("goal_topic", "/goal_pose");
    plan_topic_ = declare_parameter<std::string>("plan_topic", "/plan");
    global_frame_ = declare_parameter<std::string>("global_frame", "map");
    const auto rate = std::max(0.1, declare_parameter("planning_rate_hz", 2.0));
    declare_parameter("occupancy_threshold", 65);
    declare_parameter("allow_unknown", false);
    declare_parameter("inflation_radius_m", 0.25);
    declare_parameter("waypoint_spacing_m", 0.10);
    declare_parameter("goal_tolerance_m", 0.15);

    const auto sensor_qos = rclcpp::SensorDataQoS().keep_last(1);
    const auto map_qos = rclcpp::QoS(1).reliable().transient_local();
    plan_pub_ = create_publisher<nav_msgs::msg::Path>(plan_topic_, 10);
    map_sub_ = create_subscription<nav_msgs::msg::OccupancyGrid>(
      map_topic_, map_qos,
      [this](nav_msgs::msg::OccupancyGrid::ConstSharedPtr message) {map_ = message;});
    odom_sub_ = create_subscription<nav_msgs::msg::Odometry>(
      odom_topic_, sensor_qos,
      [this](nav_msgs::msg::Odometry::ConstSharedPtr message) {odom_ = message;});
    goal_sub_ = create_subscription<geometry_msgs::msg::PoseStamped>(
      goal_topic_, 10,
      [this](geometry_msgs::msg::PoseStamped::ConstSharedPtr message) {
        goal_ = message;
        ++goal_revision_;
      });
    timer_ = create_wall_timer(
      std::chrono::duration<double>(1.0 / rate), [this]() {update_plan();});
    RCLCPP_INFO(
      get_logger(),
      "Intro2AV C++ planner ready; implement planning_algorithm.cpp");
  }

private:
  carkit_intro2av_cpp::PlanningConfig configuration() const
  {
    carkit_intro2av_cpp::PlanningConfig config;
    config.occupancy_threshold = get_parameter("occupancy_threshold").as_int();
    config.allow_unknown = get_parameter("allow_unknown").as_bool();
    config.inflation_radius_m = get_parameter("inflation_radius_m").as_double();
    config.waypoint_spacing_m = get_parameter("waypoint_spacing_m").as_double();
    config.goal_tolerance_m = get_parameter("goal_tolerance_m").as_double();
    return config;
  }

  nav_msgs::msg::Path empty_path() const
  {
    nav_msgs::msg::Path path;
    path.header.stamp = now();
    path.header.frame_id = global_frame_;
    return path;
  }

  void update_plan()
  {
    if (!map_ || !odom_ || !goal_) {
      return;
    }
    const auto monotonic_now = std::chrono::steady_clock::now();
    if (
      planned_revision_ == goal_revision_ &&
      monotonic_now - last_plan_time_ < 1s)
    {
      return;
    }
    last_plan_time_ = monotonic_now;
    auto path = empty_path();
    try {
      auto poses = carkit_intro2av_cpp::compute_path(*map_, *odom_, *goal_, configuration());
      for (auto & pose : poses) {
        if (!std::isfinite(pose.pose.position.x) || !std::isfinite(pose.pose.position.y)) {
          throw std::runtime_error("planner returned a non-finite waypoint");
        }
        pose.header = path.header;
        path.poses.push_back(std::move(pose));
      }
    } catch (const std::exception & error) {
      RCLCPP_ERROR(get_logger(), "Planning algorithm failed: %s", error.what());
      path = empty_path();
    }
    planned_revision_ = goal_revision_;
    plan_pub_->publish(path);
  }

  std::string map_topic_, odom_topic_, goal_topic_, plan_topic_, global_frame_;
  nav_msgs::msg::OccupancyGrid::ConstSharedPtr map_;
  nav_msgs::msg::Odometry::ConstSharedPtr odom_;
  geometry_msgs::msg::PoseStamped::ConstSharedPtr goal_;
  uint64_t goal_revision_{0};
  uint64_t planned_revision_{static_cast<uint64_t>(-1)};
  std::chrono::steady_clock::time_point last_plan_time_{};
  rclcpp::Publisher<nav_msgs::msg::Path>::SharedPtr plan_pub_;
  rclcpp::Subscription<nav_msgs::msg::OccupancyGrid>::SharedPtr map_sub_;
  rclcpp::Subscription<nav_msgs::msg::Odometry>::SharedPtr odom_sub_;
  rclcpp::Subscription<geometry_msgs::msg::PoseStamped>::SharedPtr goal_sub_;
  rclcpp::TimerBase::SharedPtr timer_;
};

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<Intro2AvPlanning>());
  rclcpp::shutdown();
  return 0;
}
